### ISSUE TITLE
Mismatched carta-frontend and carta-backend versions running together

### ISSUE THREAD

**Issue by ajm-ska** [CARTAvis/carta-frontend#2216]
**Describe the bug**
It appears that incompatible carta-frontend and carta-backend versions can be run together. For example, the v3.0.0 carta-backend can use the v4.0.0 carta-frontend. The most likely scenario for this to happen is with carta-controller if the user only updates carta-controller (which updates carta-frontend) and forgets, or is unaware of the need, to update the carta-backend to the same version. At first, everything will appear to be running fine, but unexpected behaviour will occur.

It would be better if different release/beta versions of the carta-frontend and carta-backend refused to work together. 
How it could be done can be discussed. Perhaps with the carta-frontend (and/or carta-backend) can compare their version strings or protobuf versions and alert the user with an error or warning message if they do not match? 

**To Reproduce**
1. Start up a v3.0.0 carta_backend with a v4.0.0 carta-frontend using the `--frontend_folder` flag.
2. Open a HDF5 image.
3. The image will open, but no actual image will be visible.
4. Open a FITS image.
5.  The carta_backend may crash.

**Expected behavior**
Different carta-frontend and carta-backend versions should not work together. We should expect an error or warning message that the carta-frontend and carta-backend versions are incompatible.

**Screenshots or videos**
I can load a HDF5 file, but no image is visible. It is not clear to the user why, but it is because the carta_backend is v3.0.0 and the carta-frontend is v4.0.0-beta.1.
![Screenshot 2023-07-26 at 8 08 29 PM](https://github.com/CARTAvis/carta-frontend/assets/20802551/92f68789-3072-4147-b33b-90c2f85a7e37)

**Platform info (please complete the following information):**
 - OS [e.g. macOS Monterey]: macOS Ventura
 - Browser [e.g. chrome, safari, electron app]: Chrome 
 - Browser version [e.g. 22]: 112
 - Backend branch [e.g. dev, v3b2 release]: v3.0.0
 - Frontend branch [e.g. dev, v3b2 release]: v4.0.0-beta.1

**Additional context**
I will make an identical carta-backend issue as I'm not sure if the carta-backend would also need modification to address this issue.

**Comment by ajm-ska** [comment]
Also posted as [carta-backend issue 1281](https://github.com/CARTAvis/carta-backend/issues/1281)

**Comment by kswang1029** [comment]
Or maybe to the controller repo which may be enhanced by forcing updating both frontend and backend if controller is being updated?

**Comment by confluence** [comment]
The controller has no control over the installation of the backend, which is managed independently.

Currently the backend checks for protobuf interface compatibility when registering the viewer (using an interface version number sent by the frontend). For consistency, any additional backend/frontend version compatibility check should probably also go in the backend, so that all the heuristics for accepting or rejecting the connection can go in one place. So this has a protobuf dependency, because a frontend version should be added to the `RegisterViewer` message.

**Comment by veggiesaurus** [comment]
> It would be better if different release/beta versions of the carta-frontend and carta-backend refused to work together. How it could be done can be discussed. Perhaps with the carta-frontend (and/or carta-backend) can compare their version strings or protobuf versions and alert the user with an error or warning message if they do not match?

I am still convinced that we have everything we need to do exactly this in the ICD version number. We just have not been good at all about bumping it on breaking changes. We should be bumping the ICD version number whenever a breaking change occurs. The fact that we get a crash means that (a) we didn't change the ICD version number or (b) we aren't displaying sufficient warning on the frontend.

Surely introducing _another_ more complex field (an array of frontend versions) is going to lead to even more situations where we don't keep things updated correctly?

**Comment by markccchiang** [comment]
> > It would be better if different release/beta versions of the carta-frontend and carta-backend refused to work together. How it could be done can be discussed. Perhaps with the carta-frontend (and/or carta-backend) can compare their version strings or protobuf versions and alert the user with an error or warning message if they do not match?
> 
> I am still convinced that we have everything we need to do exactly this in the ICD version number. We just have not been good at all about bumping it on breaking changes. We should be bumping the ICD version number whenever a breaking change occurs. The fact that we get a crash means that (a) we didn't change the ICD version number or (b) we aren't displaying sufficient warning on the frontend.
> 
> Surely introducing _another_ more complex field (an array of frontend versions) is going to lead to even more situations where we don't keep things updated correctly?

I agree with @veggiesaurus that checking ICD version number is a simpler way. Just we need to remember to update it whenever the key changes in the code. @confluence and @kswang1029 do you have other thoughts or concerns?

**Comment by kswang1029** [comment]
Matching ICD version is good to ensure protobuf message integrity. We might need to consider the case that the frontend uses the data in the protobuf message _differently_. If this is the case, we would need extra info such as frontend/backend release version string, for example.

**Comment by veggiesaurus** [comment]
> Matching ICD version is good to ensure protobuf message integrity. We might need to consider the case that the frontend uses the data in the protobuf message _differently_. If this is the case, we would need extra info such as frontend/backend release version string, for example. 

The icd version should represent more than the protbuf message integrity. In theory it should represent the contract between frontend and backend. If we alter that contract, a new version number should be added

**Comment by veggiesaurus** [comment]
(PS I realise I am the one most responsible for breaking the above rule 🙈🙈🙈)

**Comment by kswang1029** [comment]
> > Matching ICD version is good to ensure protobuf message integrity. We might need to consider the case that the frontend uses the data in the protobuf message _differently_. If this is the case, we would need extra info such as frontend/backend release version string, for example.
> 
> The icd version should represent more than the protbuf message integrity. In theory it should represent the contract between frontend and backend. If we alter that contract, a new version number should be added

Consider the following example: currently we display per-channel histogram in the render config widget. If we now add a new feature by allowing users to define a region to display its region histogram in the render config widget (likely what we need to do is to re-use region histogram protobuf messages and add new logics in the frontend codebase), do you mean in this case we also have to bump the ICD version too? If this is handled rigorously from now on, yes, a single ICD version string should do the job.

**Comment by acdo2002** [comment]
The current ICD repository should bump the carta-protobuf submodule together with the frontend/backend/protobuf.
If adding a single ICD version in the config.json to make sure the ICD version, it should be fine.

**Comment by veggiesaurus** [comment]
> > > Matching ICD version is good to ensure protobuf message integrity. We might need to consider the case that the frontend uses the data in the protobuf message _differently_. If this is the case, we would need extra info such as frontend/backend release version string, for example.
> > 
> > 
> > The icd version should represent more than the protbuf message integrity. In theory it should represent the contract between frontend and backend. If we alter that contract, a new version number should be added
> 
> Consider the following example: currently we display per-channel histogram in the render config widget. If we now add a new feature by allowing users to define a region to display its region histogram in the render config widget (likely what we need to do is to re-use region histogram protobuf messages and add new logics in the frontend codebase), do you mean in this case we also have to bump the ICD version too? If this is handled rigorously from now on, yes, a single ICD version string should do the job.

I guess that depends on implementation. If we're re-using an existing set of functionality on the backend, then surely the older version of the backend would be able to handle it? If we're implementing new functionality on the backend that previously wasn't supported, then we would need to bump the version number

**Comment by markccchiang** [comment]
> > Consider the following example: currently we display per-channel histogram in the render config widget. If we now add a new feature by allowing users to define a region to display its region histogram in the render config widget (likely what we need to do is to re-use region histogram protobuf messages and add new logics in the frontend codebase), do you mean in this case we also have to bump the ICD version too? If this is handled rigorously from now on, yes, a single ICD version string should do the job.
> 
> I guess that depends on implementation. If we're re-using an existing set of functionality on the backend, then surely the older version of the backend would be able to handle it? If we're implementing new functionality on the backend that previously wasn't supported, then we would need to bump the version number

In such a case, we expand the meaning of ICD version that is not only for the protobuf message but also includes the functionalities for the backend/frontend. Should we redefine the ICD version number so that it is easy to be identified? For example, the current backend/frontend version is `4.0.0`, so the ICD version number is `400`? The range of uint16_t is 0 ~ 65,535. Ps: [software versioning](https://en.wikipedia.org/wiki/Software_versioning).

**Comment by kswang1029** [comment]
Just a thought: if changing version string depends on implementation case by case (meaning we need someone to keep an eye on this and we may make mistakes), I would suggest we keep the current way and just use a new build tag for the matching purpose. _After all, this matching issue only matters when releasing CARTA to users._ The new build tag by default can be as "dev" so we have freedom to mix n match backend and frontend for testing purpose. When we create release packages, we change to the build flag to something else for matching.
